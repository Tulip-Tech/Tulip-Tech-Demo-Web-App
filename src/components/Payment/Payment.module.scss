@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';

.infoBox {
  display: flex;
  justify-content: space-between;
  margin-bottom: variables.$base-spacing;
  padding: calc(#{variables.$base-spacing} / 2) variables.$base-spacing;

  font-size: 14px;
  line-height: 18px;
  text-shadow: 0 2px 4px rgba(0, 0, 0, 0.14), 0 3px 4px rgba(0, 0, 0, 0.12), 0 1px 5px rgba(0, 0, 0, 0.2);
  background: theme.$panel-bg;
  border-radius: 4px;

  > strong {
    line-height: 16px;
    letter-spacing: 0.25px;
  }
}

.price {
  font-size: 14px;
  line-height: 18px;
  > strong {
    font-weight: var(--body-font-weight-bold);
    font-size: 24px;
    line-height: 26px;
  }
}

.cardDetails {
  display: flex;
  margin-top: variables.$base-spacing * 2;
}
