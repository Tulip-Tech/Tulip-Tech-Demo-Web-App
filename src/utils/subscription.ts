import { formatPrice } from '#src/utils/formatting';
import type { Offer, Order } from '#types/checkout';

export const getOfferPrice = (offer: Offer) => formatPrice(offer.customerPriceInclTax, offer.customerCurrency, offer.customerCountry);

export const isSVODOffer = (offer: Offer | Order) => offer.offerId?.[0] === 'S';
