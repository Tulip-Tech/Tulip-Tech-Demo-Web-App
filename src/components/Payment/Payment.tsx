import React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';

import styles from './Payment.module.scss';

import TextField from '#components/TextField/TextField';
import LoadingOverlay from '#components/LoadingOverlay/LoadingOverlay';
import Button from '#components/Button/Button';
import type { Customer } from '#types/account';
import { formatDate, formatPrice } from '#src/utils/formatting';
import { addQueryParam } from '#src/utils/location';
import type { PaymentDetail, Subscription, Transaction } from '#types/subscription';
import type { AccessModel } from '#types/Config';

const VISIBLE_TRANSACTIONS = 4;

type Props = {
  accessModel: AccessModel;
  activeSubscription: Subscription | null;
  activePaymentDetail: PaymentDetail | null;
  transactions: Transaction[] | null;
  customer: Customer;
  isLoading: boolean;
  panelClassName?: string;
  panelHeaderClassName?: string;
  onShowAllTransactionsClick?: () => void;
  showAllTransactions: boolean;
  canRenewSubscription?: boolean;
};

const Payment = ({
  accessModel,
  activePaymentDetail,
  activeSubscription,
  transactions,
  customer,
  isLoading,
  panelClassName,
  panelHeaderClassName,
  onShowAllTransactionsClick,
  showAllTransactions,
  canRenewSubscription = false,
}: Props): JSX.Element => {
  const { t } = useTranslation(['user', 'account']);
  const hiddenTransactionsCount = transactions ? transactions?.length - VISIBLE_TRANSACTIONS : 0;
  const hasMoreTransactions = hiddenTransactionsCount > 0;
  const navigate = useNavigate();
  const location = useLocation();

  function onCompleteSubscriptionClick() {
    navigate(addQueryParam(location, 'u', 'choose-offer'));
  }

  function onCancelSubscriptionClick() {
    navigate(addQueryParam(location, 'u', 'unsubscribe'));
  }

  function onRenewSubscriptionClick() {
    navigate(addQueryParam(location, 'u', 'renew-subscription'));
  }

  function getTitle(period: Subscription['period']) {
    switch (period) {
      case 'month':
        return t('user:payment.monthly_subscription');
      case 'year':
        return t('user:payment.annual_subscription');
      case 'day':
        return t('user:payment.daily_subscription');
      case 'week':
        return t('user:payment.weekly_subscription');
      case 'granted':
        return t('user:payment.granted_subscription');
      default:
        return t('user:payment.other');
    }
  }

  return (
    <>
      {accessModel === 'SVOD' && (
        <div className={panelClassName}>
          <div className={panelHeaderClassName}>
            <h3>{t('user:payment.subscription_details')}</h3>
          </div>
          {activeSubscription ? (
            <React.Fragment>
              <div className={styles.infoBox} key={activeSubscription.subscriptionId}>
                <p>
                  <strong>{getTitle(activeSubscription.period)}</strong> <br />
                  {activeSubscription.status === 'active' && activeSubscription.period !== 'granted'
                    ? t('user:payment.next_billing_date_on', { date: formatDate(activeSubscription.expiresAt) })
                    : t('user:payment.subscription_expires_on', { date: formatDate(activeSubscription.expiresAt) })}
                </p>
                <p className={styles.price}>
                  <strong>{formatPrice(activeSubscription.nextPaymentPrice, activeSubscription.nextPaymentCurrency, customer.country)}</strong>
                  <small>/{t(`account:periods.${activeSubscription.period}`)}</small>
                </p>
              </div>
              {activeSubscription.status === 'active' && activeSubscription.period !== 'granted' ? (
                <Button label={t('user:payment.cancel_subscription')} onClick={onCancelSubscriptionClick} />
              ) : canRenewSubscription ? (
                <Button label={t('user:payment.renew_subscription')} onClick={onRenewSubscriptionClick} />
              ) : null}
            </React.Fragment>
          ) : isLoading ? null : (
            <React.Fragment>
              <p>{t('user:payment.no_subscription')}</p>
              <Button variant="contained" color="primary" label={t('user:payment.complete_subscription')} onClick={onCompleteSubscriptionClick} />
            </React.Fragment>
          )}
        </div>
      )}
      <div className={panelClassName}>
        <div className={panelHeaderClassName}>
          <h3>{t('user:payment.payment_method')}</h3>
        </div>
        {activePaymentDetail ? (
          <div key={activePaymentDetail.id}>
            <TextField
              label={t('user:payment.card_number')}
              value={`•••• •••• •••• ${activePaymentDetail.paymentMethodSpecificParams.lastCardFourDigits || ''}`}
              editing={false}
            />
            <div className={styles.cardDetails}>
              <TextField label={t('user:payment.expiry_date')} value={activePaymentDetail.paymentMethodSpecificParams.cardExpirationDate} editing={false} />
              <TextField label={t('user:payment.cvc_cvv')} value={'******'} editing={false} />
            </div>
          </div>
        ) : (
          <div>
            <p>{!isLoading && t('user:payment.no_payment_methods')}</p>
          </div>
        )}
      </div>
      <div className={panelClassName}>
        <div className={panelHeaderClassName}>
          <h3>{t('user:payment.transactions')}</h3>
        </div>
        {transactions?.length ? (
          <React.Fragment>
            {transactions?.slice(0, showAllTransactions ? 9999 : VISIBLE_TRANSACTIONS).map((transaction) => (
              <div className={styles.infoBox} key={transaction.transactionId}>
                <p>
                  <strong>{transaction.offerTitle}</strong> <br />
                  {t('user:payment.price_payed_with', {
                    price: formatPrice(parseInt(transaction.transactionPriceInclTax), transaction.transactionCurrency, transaction.customerCountry),
                    method: transaction.paymentMethod,
                  })}
                </p>
                <p>
                  {transaction.transactionId}
                  <br />
                  {formatDate(transaction.transactionDate)}
                </p>
              </div>
            ))}
            {!showAllTransactions && hasMoreTransactions ? (
              <React.Fragment>
                <p>{t('user:payment.hidden_transactions', { count: hiddenTransactionsCount })}</p>
                <Button label={t('user:payment.show_all')} onClick={onShowAllTransactionsClick} />
              </React.Fragment>
            ) : null}
          </React.Fragment>
        ) : (
          <div>
            <p>{!isLoading && t('user:payment.no_transactions')}</p>
          </div>
        )}
      </div>
      {isLoading && <LoadingOverlay inline />}
    </>
  );
};

export default Payment;
