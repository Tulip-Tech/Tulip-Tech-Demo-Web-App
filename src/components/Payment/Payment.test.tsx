import React from 'react';

import Payment from './Payment';

import customer from '#test/fixtures/customer.json';
import transactions from '#test/fixtures/transactions.json';
import paymentDetail from '#test/fixtures/paymentDetail.json';
import subscription from '#test/fixtures/subscription.json';
import type { Customer } from '#types/account';
import type { PaymentDetail, Subscription, Transaction } from '#types/subscription';
import { renderWithRouter } from '#test/testUtils';

describe('<Payment>', () => {
  test('renders and matches snapshot', () => {
    const { container } = renderWithRouter(
      <Payment
        accessModel="AVOD"
        customer={customer as Customer}
        transactions={transactions as Transaction[]}
        activeSubscription={subscription as Subscription}
        activePaymentDetail={paymentDetail as PaymentDetail}
        showAllTransactions={false}
        isLoading={false}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
