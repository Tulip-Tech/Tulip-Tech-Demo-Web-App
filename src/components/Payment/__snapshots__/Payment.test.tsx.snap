// Vitest Snapshot v1

exports[`<Payment> > renders and matches snapshot 1`] = `
<div>
  <div>
    <div>
      <h3>
        user:payment.payment_method
      </h3>
    </div>
    <div>
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235"
        >
          user:payment.card_number
        </label>
        <p>
          •••• •••• •••• 0002
        </p>
      </div>
      <div
        class="_cardDetails_c57ff5"
      >
        <div
          class="_textField_e16c1b"
        >
          <label
            class="_label_e16c1b"
            for="text-field_1235"
          >
            user:payment.expiry_date
          </label>
          <p>
            03/2030
          </p>
        </div>
        <div
          class="_textField_e16c1b"
        >
          <label
            class="_label_e16c1b"
            for="text-field_1235"
          >
            user:payment.cvc_cvv
          </label>
          <p>
            ******
          </p>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <h3>
        user:payment.transactions
      </h3>
    </div>
    <div
      class="_infoBox_c57ff5"
    >
      <p>
        <strong>
          Monthly subscription (recurring)
        </strong>
         
        <br />
        user:payment.price_payed_with
      </p>
      <p>
        T712014024
        <br />
        5/5/2021
      </p>
    </div>
    <div
      class="_infoBox_c57ff5"
    >
      <p>
        <strong>
          Monthly subscription (recurring)
        </strong>
         
        <br />
        user:payment.price_payed_with
      </p>
      <p>
        T177974068
        <br />
        5/5/2021
      </p>
    </div>
    <div
      class="_infoBox_c57ff5"
    >
      <p>
        <strong>
          Monthly subscription (recurring)
        </strong>
         
        <br />
        user:payment.price_payed_with
      </p>
      <p>
        T996601696
        <br />
        5/5/2021
      </p>
    </div>
  </div>
</div>
`;
